File: common/src/main/java/com/pluxity/file/config/SseConfig.java
package com.pluxity.file.config;

import jakarta.annotation.Nonnull;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.servlet.AsyncHandlerInterceptor;
import org.springframework.web.servlet.config.annotation.AsyncSupportConfigurer;
import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

@Configuration
public class SseConfig implements WebMvcConfigurer {
    /** 모든 비동기 요청에 대한 기본 타임아웃을 설정합니다. 이는 SSE가 아닌 다른 비동기 요청에 대한 안전장치 역할을 합니다. */
    @Override
    public void configureAsyncSupport(AsyncSupportConfigurer configurer) {
        configurer.setDefaultTimeout(30_000L); // 30 seconds
    }

    /** 커스텀 인터셉터를 등록합니다. 이 인터셉터는 SseEmitter를 반환하는 요청을 식별하여 타임아웃을 재설정합니다. */
    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(new SseTimeoutSettingInterceptor());
    }

    /** SseEmitter를 반환하는 컨트롤러 메소드에 대해 선택적으로 비동기 요청 타임아웃을 재설정하는 인터셉터입니다. */
    static class SseTimeoutSettingInterceptor implements AsyncHandlerInterceptor {

        // 1시간 타임아웃 (밀리초 단위)
        private static final long SSE_TIMEOUT = 60 * 60 * 1000L;

        /**
         * 컨트롤러 메소드가 실행되기 전에 호출됩니다.
         *
         * @param handler 처리될 엔드포인트의 핸들러 (컨트롤러 메소드 정보 포함)
         * @return 요청 처리를 계속하려면 true, 중단하려면 false
         */
        @Override
        public boolean preHandle(
                @Nonnull HttpServletRequest request,
                @Nonnull HttpServletResponse response,
                @Nonnull Object handler) {
            // 핸들러가 실제 컨트롤러 메소드를 나타내는 HandlerMethod 인스턴스인지 확인합니다.
            if (handler instanceof HandlerMethod handlerMethod) {
                // 해당 컨트롤러 메소드의 반환 타입이 SseEmitter 클래스이거나 그 자식 클래스인지 확인합니다.
                if (SseEmitter.class.isAssignableFrom(handlerMethod.getReturnType().getParameterType())) {
                    // 조건이 일치하면, 이 특정 요청에 대한 서블릿 컨테이너의 비동기 타임아웃을
                    // SSE를 위한 긴 타임아웃 값으로 덮어씁니다.
                    request.getAsyncContext().setTimeout(SSE_TIMEOUT);
                }
            }
            // 모든 경우에 요청 처리를 계속 진행하도록 true를 반환합니다.
            return true;
        }
    }
}

File: common/src/main/kotlin/com/pluxity/file/config/SseConfig.kt
package com.pluxity.file.config

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.context.annotation.Configuration
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.AsyncHandlerInterceptor
import org.springframework.web.servlet.config.annotation.AsyncSupportConfigurer
import org.springframework.web.servlet.config.annotation.InterceptorRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter

@Configuration
class SseConfig : WebMvcConfigurer {
    override fun configureAsyncSupport(configurer: AsyncSupportConfigurer) {
        // 기본 타임아웃은 여전히 30초로 유지
        configurer.setDefaultTimeout(30000L)
    }

    override fun addInterceptors(registry: InterceptorRegistry) {
        // 이 인터셉터는 이제 afterConcurrentHandlingStarted를 사용하므로
        // preHandle에서는 아무 일도 하지 않습니다.
        registry.addInterceptor(SseTimeoutSettingInterceptor())
    }

    /** SseEmitter를 반환하여 비동기 요청 처리가 시작된 후, 해당 요청의 타임아웃을 재설정하는 인터셉터.  */
    internal class SseTimeoutSettingInterceptor : AsyncHandlerInterceptor {
        /** 이 메소드는 컨트롤러가 SseEmitter를 반환하고, 서블릿 스레드가 풀에 반환된 직후(즉, 비동기 처리가 시작된 후)에 호출됩니다.  */
        override fun afterConcurrentHandlingStarted(
            request: HttpServletRequest,
            response: HttpServletResponse,
            handler: Any,
        ) {
            // preHandle에서 했던 것과 동일한 검증 로직 수행
            if (handler is HandlerMethod) {
                if (SseEmitter::class.java.isAssignableFrom(handler.returnType.getParameterType())) {
                    // 이제 isAsyncStarted()가 true인 안전한 시점이므로, getAsyncContext() 호출 가능
                    request.asyncContext.timeout = SSE_TIMEOUT
                }
            }
        }

        companion object {
            private const val SSE_TIMEOUT = 30 * 60 * 1000L // 30분
        }
    }
}
